### ISSUE TITLE
React-dom and Apple TV

### ISSUE THREAD

**Issue by nickdima** [facebook/react#5439]
I've been playing lately with getting React to work for building client/server apps for the new Apple TV so I'm opening this thread just to start the conversation around the issue in case other people are also looking into it.

For those unfamiliar with the subject the SDK for the new Apple TV lets you build JavaScript applications that get loaded from your server and evaluated in a JS context on the device. Now for building interfaces they came up with a [DOM implementation](https://developer.apple.com/library/prerelease/tvos/documentation/TVMLJS/Reference/TVJSFrameworkReference/) and some flavour of XML called [TVML](https://developer.apple.com/library/prerelease/tvos/documentation/LanguagesUtilities/Conceptual/ATV_Template_Guide/) that just bridges to native UI components and APIs (much like React Native in a way). They also have a small JavaScript API called [TVJS](https://developer.apple.com/library/prerelease/tvos/documentation/TVMLJS/Reference/TVJSFrameworkReference/) for creating documents, making HTTP calls, etc.

This way of building apps is an alternative to the traditional iOS development that you can also do on the Apple TV with Swift/Obj-C, UIKit and the other frameworks. The interesting thing is that the default Apple TV apps like Music, TV Shows, etc. are built as JS client/server apps with Browserify and Handlebars templates.

So at a first try React kind of works on the platform (you might need to shim `window` and `document` as they are not available) but as you start getting into the details you realise there are some limitations to ATV's DOM model. For example I just noticed that [IDL attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes#Content_versus_IDL_attributes) are not working so that might be a problem. Then some elements are conflicting with HTML elements that have the same name, same issue for property names, and so on.

In conclusion I think a custom React renderer for the Apple TV would be the solution but the strange thing is that it would still be a DOM renderer like `react-dom` only configured for ATV's elements and attributes and patched for its DOM implementation's quirks. Do you guys think this is feasible?

**Comment by jimfb** [comment]
Yes, totally feasible.  That's the reason that we split react-dom from react-core in 0.14 - to allow custom renderers like this to be written easily.  The best approach is probably to fork react-dom, and build on top of that.  From your description, it sounds like `src/renderers/dom/shared/HTMLDOMPropertyConfig.js` would be a good place for you to start making your changes.

It sounds like a usage question rather than a bug in the React core.  I'm going to close this out, since we use github issues to track bugs in the React core.  Usage questions are better addressed on sites like StackOverflow.  Feel free to continue the discussion on this thread.

**Comment by nickdima** [comment]
Thanks @jimfb! I started working on the fork, `HTMLDOMPropertyConfig` was a good starting point.
By looking at `react-dom` it looks like there's a lot of code related to html tags, css attributes, browser differences, etc. that won't be needed in this fork, so probably would be good do remove this parts. Do you think this will cause problems later on when pulling changes from upstream?

**Comment by jimfb** [comment]
@nickdima My recommendation is that you always remain as close to the original as possible when forking a project.  Always push any changes that make sense to upstream.  Otherwise, you'll likely fall behind.  That said, ultimately, you need to make whatever changes are required for your rendering target.

**Comment by nickdima** [comment]
Do you think it would make sense for this to be part of `react-dom`? I mean the target is still a DOM implementation in the end.
Just looking for the best solution as I'm sure this renderer could be useful.

**Comment by jimfb** [comment]
I don't know enough about the differences to have an opinion on that.  I suspect that a bunch of your issues may be solved by #140 (actually, that might solve nearly everything for you), but I just don't have enough context to make an educated statement on TVML.

**Comment by cezary** [comment]
@nickdima have you seen [react-tvml](https://github.com/ramitos/react-tvml)?

**Comment by nickdima** [comment]
Yeah, just saw it a couple of days ago after implementing more or less the same thing.
Unfortunately this is the only possible solution. It's not so elegant because a lot of code from ReactDOM needs to be replicated so it could be difficult keeping it synced with the original code.

**Comment by antigirl** [comment]
Have you had a look again since reactDom has been abstracted out of React core?
